src/data/KeyspaceBackend.hpp

```diff
@@ -0,0 +1,308 @@
+#include "data/cassandra/SettingsProvider.hpp"
+#include "data/cassandra/Types.hpp"
+#include "data/cassandra/impl/ExecutionStrategy.hpp"
+#include "util/Assert.hpp"
+#include "util/log/Logger.hpp"
+
+#include <boost/asio/spawn.hpp>
+#include <boost/json/object.hpp>
+#include <boost/uuid/string_generator.hpp>
+#include <boost/uuid/uuid.hpp>
+#include <cassandra.h>
+#include <fmt/format.h>
+#include <xrpl/basics/Blob.h>
+#include <xrpl/basics/base_uint.h>
+#include <xrpl/basics/strHex.h>
+#include <xrpl/protocol/AccountID.h>
+#include <xrpl/protocol/Indexes.h>
+#include <xrpl/protocol/LedgerHeader.h>
+#include <xrpl/protocol/nft.h>
+
+#include <cstddef>
+#include <cstdint>
+#include <iterator>
+#include <optional>
+#include <stdexcept>
+#include <utility>
+#include <vector>
+
+namespace data::cassandra {
+
+/**
+ * @brief Implements @ref CassandraBackendFamily for Keyspace
+ *
+ * @tparam SettingsProviderType The settings provider type to use
+ * @tparam ExecutionStrategyType The execution strategy type to use
+ * @tparam FetchLedgerCacheType The ledger header cache type to use
+ */
+template <
+    SomeSettingsProvider SettingsProviderType,
+    SomeExecutionStrategy ExecutionStrategyType,
+    typename FetchLedgerCacheType = FetchLedgerCache>
+class BasicKeyspaceBackend : public CassandraBackendFamily<
+                                 SettingsProviderType,
+                                 ExecutionStrategyType,
+                                 KeyspaceSchema<SettingsProviderType>,
+                                 FetchLedgerCacheType> {
+    using DefaultCassandraFamily = CassandraBackendFamily<
+        SettingsProviderType,
+        ExecutionStrategyType,
+        KeyspaceSchema<SettingsProviderType>,
+        FetchLedgerCacheType>;
+
+    using DefaultCassandraFamily::executor_;
+    using DefaultCassandraFamily::ledgerSequence_;
+    using DefaultCassandraFamily::log_;
+    using DefaultCassandraFamily::range_;
+    using DefaultCassandraFamily::schema_;
+
+public:
+    /**
+     * @brief Inherit the constructors of the base class.
+     */
+    using DefaultCassandraFamily::DefaultCassandraFamily;
+
+    /**
+     * @brief Move constructor is deleted because handle_ is shared by reference with executor
+     */
+    BasicKeyspaceBackend(BasicKeyspaceBackend&&) = delete;
+
+    bool
+    doFinishWrites() override
+    {
+        this->waitForWritesToFinish();
+
+        // !range_.has_value() means the table 'ledger_range' is not populated;
+        // This would be the first write to the table.
+        // In this case, insert both min_sequence/max_sequence range into the table.
+        if (not(range_.has_value())) {
+            executor_.writeSync(schema_->insertLedgerRange, false, ledgerSequence_);
+            executor_.writeSync(schema_->insertLedgerRange, true, ledgerSequence_);
+        }
+
+        if (not this->executeSyncUpdate(schema_->updateLedgerRange.bind(ledgerSequence_, true, ledgerSequence_ - 1))) {
```

Comment:
I would probably add inline comments to a boolean variable like `executor_.writeSync(schema_->insertLedgerRange, /* some_opt_name = */ false, ledgerSequence_);`